### PR TITLE
Bug 1283930 - Add Makefile.PL & local/lib/perl5 support to bmo/master…

### PR DESCRIPTION
**Pull request by dylanwh**
… + local symlink to data/ directory

This changes the copyright notices, ensures use 5.10.1 / use strict / use warnings
and adds the use lib (... local/lib/perl5). It also adds a "local" symlink
and updates a number of tests so they pass. It also have the version of vers_cmp() in Bugzilla::Version.
I've tested pulling other patches out of upstream-merge and that works about as well as you could expect. If we can land this the next thing I'd pull is Cache::Memcached::Fast related code.

### CODE REVIEW

**Comment by moz-hwine** [comment]
(Automated Close) Please do not file PRs here, see https://wiki.mozilla.org/BMO/Submitting_Patches

File: Bugzilla/Install/Requirements.pm [hunk 1]
```diff
@@ -81,471 +72,140 @@ use constant APACHE => qw(apachectl httpd apache2 apache);
...
+        # detaint
+        $str =~ /^(.+)$/s; $str = $1;
+        close $meta_fh;
+
+        return CPAN::Meta->load_json_string($str);
     }
-    return \@modules;
-};
+    else {
+        ThrowCodeError('cpan_meta_missing');
```
Comment (dklawren) [inline Bugzilla/Install/Requirements.pm]: Need to load Bugzilla::Error at top otherwise it dies not able to find ThrowCodeError()

**Comment by dklawren** [comment]
- Bugzilla::Error needs to be included in Bugzilla::Install::Requirements
- When running perl t/001compile.t before running perl Makefile.PL it fails with the following:

$ perl t/001compile.t 
1..684
ok 1 - use Bugzilla::Constants;
ok 2 - use Bugzilla::Install::Requirements;
ok 3 - use Bugzilla;
ok 4 - admin.cgi
ok 5 - attachment.cgi
$VAR1 = {
          'error' => 'cpan_meta_missing'
        };
# Looks like you planned 684 tests but ran 5.
# Looks like your test exited with 255 just after 5.

This does not happen on upstream/master

- When running perl Makefile.PL I get a lot of redefined warnings:

$ perl Makefile.PL 
\Hiding META.{json,yml} from MakeMaker...
Writing Makefile for Bugzilla
Writing MYMETA.yml and MYMETA.json
Constant subroutine Bugzilla::Extension::AntiSpam::REQUIRED_MODULES redefined at /usr/share/perl5/vendor_perl/constant.pm line 140.
Constant subroutine Bugzilla::Extension::AntiSpam::OPTIONAL_MODULES redefined at /usr/share/perl5/vendor_perl/constant.pm line 140.
Subroutine REQUIRED_MODULES redefined at ./extensions/Bitly/Config.pm line 16.
Constant subroutine Bugzilla::Extension::Bitly::OPTIONAL_MODULES redefined at /usr/share/perl5/vendor_perl/constant.pm line 140.
Constant subroutine Bugzilla::Extension::Bitly::API_VERSION_MAP redefined at /usr/share/perl5/vendor_perl/constant.pm line 140.
Constant subroutine Bugzilla::Extension::BMO::REQUIRED_MODULES redefined at /usr/share/perl5/vendor_perl/constant.pm line 140.
Constant subroutine Bugzilla::Extension::BMO::OPTIONAL_MODULES redefined at /usr/share/perl5/vendor_perl/constant.pm line 140.
Constant subroutine Bugzilla::Extension::BugmailFilter::REQUIRED_MODULES redefined at /usr/share/perl5/vendor_perl/constant.pm line 140.
Constant subroutine Bugzilla::Extension::BugmailFilter::OPTIONAL_MODULES redefined at /usr/share/perl5/vendor_perl/constant.pm line 140.
...
Unhiding META.{json,yml}...

- MYMETA.{json,yml} is included in the pull request but not META.{json,yml}

File: Bugzilla/Install/Requirements.pm [hunk 2]
```diff
@@ -141,15 +141,6 @@ sub check_all_cpan_features {
...
 
-    print install_string('all_optional_features_require'), "\n" if $output;
-    my $features = check_cpan_feature($meta->feature('features'), $dirs, $output);
-    print "\n" if $output;
-
-    $report{features} = {
-        description => $meta->feature('features')->description,
-        result => $features,
-    };
-
```
Comment (dklawren) [inline Bugzilla/Install/Requirements.pm]: What was the reason for removing this section I am curious?

File: Makefile.PL [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+            foreach my $feature (@features) {
+                $optional_features{$feature}{prereqs}{runtime}{requires}
+                    { $mod->{module} } = $mod->{version} // 0;
+            }
+        }
+    }
+}
+
+# BMO Customization
+my @bmo_features = grep { !/pg|oracle|mod_perl|sqlite/i } keys %optional_features;
```
Comment (dklawren) [inline Makefile.PL]: I understand why you added this but if I do --with-feature bmo I also get lots of other optional features enabled that are not used on BMO.

Change to:
+my @bmo_features = grep { !/pg|oracle|mod_perl|sqlite|auth_ldap|auth_radius|smtp_auth|updates/i } keys %optional_features;

**Comment by dklawren** [comment]
I have tried installing manually LWP::Protocol::https before and after running 'cpanm --installdeps .' and it works fine. But installation of that module fails when installed as part of 'cpanm --installdeps .' so we may need to find a work around for that.

**Pull request closed without merging**